src/main-fuse.h

```diff
@@ -13,7 +13,11 @@ int hfs_open(const char* path, struct fuse_file_info* info);
 int hfs_read(const char* path, char* buf, size_t bytes, off_t offset, struct fuse_file_info* info);
 int hfs_release(const char* path, struct fuse_file_info* info);
 int hfs_readdir(const char* path, void* buf, fuse_fill_dir_t filler, off_t offset, struct fuse_file_info* info);
-int hfs_getxattr(const char* path, const char* name, char* value, size_t vlen);
+#ifdef __APPLE__
```

Comment:
We have to build non-`__APPLE__` version on Darling, because we use Linux FUSE, not OSXFUSE. I.e., use `#if defined(__APPLE__) && !defined(DARLING)` or some other way to detect if OSXFUSE or Linux FUSE is in use.

Reply:
If you haven't noticed, while darling-dmg can be used standalone as a DMG FUSE implementation, it's a part of the much larger [Darling project](https://darlinghq.org/), where we use it as an implementation of `hdiutil`.

> Can you use Linux FUSE on Osx ? I guess not.

Not on macOS itself, but on Darling, yes.

> So what's wrong with ifdef APPLE?

`__APPLE__` *is* defined under Darling, because for _most_ purposes, Darling is a like macOS, but there are exceptions. This is one of them: here, we want the compiler to use the Linux declaration.

Reply:
> I looked in fuse.h and they are using #ifdef __APPLE__.

FUSE ("Linux FUSE") and OSXFUSE are _not_ the same library built for different OSes, they're different projects. OSXFUSE is intentionally basing their API on FUSE, to make it easier to port existing filesystems from Linux, but it doesn't mean the API is exactly the same (as the need for this PR demonstrates). The `#ifdef __APPLE__`-s [aren't in the original FUSE](https://github.com/libfuse/libfuse/blob/master/include/fuse.h#L517), which we use, they're OSXFUSE additions.

> What happen if __APPLE__ and DARLING is defined ? Fuse will expect one more arg, but getxattr from main-fuse won't provide it.

It won't. `DARLING` is defined when we build darling-dmg as a part of Darling, which means we're using the original, Linux version of `fuse.h`. It expects 4 args and will get 4, just as it works right now without this change.

> Ot is it that darling define __APPLE__ for darling source but not when including fuse.h ?

No, Darling always defines `__APPLE__` (maybe there are some exceptions, but this isn't one of them).

Reply:
Ok, I understand that I stupidly assumed that the headers would be the same to enforce the API, with the one exception for getxattr. Because it's enclose by `__APPLE__ `, I really though without double checking that this fuse.h would be shared.
Thanks very much for your time to explain me that.

What about `#if defined(__APPLE__) && defined(__MACH__)` ? `__MACH__` should exclude all non-darwin build. Seems to be standard (https://sourceforge.net/p/predef/wiki/OperatingSystems/)

Reply:
> Ot is it that darling define APPLE for darling source but not when including fuse.h ?

But don't get me wrong, sometimes we have to go out of our way to convince the headers we include that we're Linux (or, to be more correct, that we want the Linux version of the header), despite us looking every bit like macOS. See e.g. [this commit](https://github.com/darlinghq/darling-cocotron/commit/5c69ad493dd188d6874dc0e5ddce7dd001c5c8e1), where we had to define `__linux__` before including some headers and then undefine it back.

It's just that the original `fuse.h` happens not to care. Otherwise, yeah, we'd have to do that for it as well.

> What about #if defined(__APPLE__) && defined(__MACH__) ? __MACH__ should exclude all non-darwin build. Seems to be standard (https://sourceforge.net/p/predef/wiki/OperatingSystems/)

Well, we have Mach and define `__MACH__` (and everything else macOS defines). You see, Darling *is* a Darwin system. It's more Darwin than even Linux (we use Darwin's libSystem and all...), but it's kinda Linux, too (it's a very _interesting_ mix, go figure).

There's no better way to check for Darling than, well, `DARLING` (but note it's only defined when building Darling itself; a program built on macOS would have to do other tricks like checking if `/usr/lib/darling` exists to figure out if it's running on Darling).

If you look at this all another way, as if darling-dmg was a third-party project, then yes, Darling cannot run macOS programs that use OSXFUSE (unless we implement a way to trick them into using a FUSE wrapper exposing OSXFUSE ABI instead...) Of course, we won't ask every macOS filesystem project to add a check for `DARLING` (and then again it won't help because they'd need a runtime check).

But darling-dmg "just so happens" to be affiliated with us Darling developers, and we're also building it as a part of Darling, so we can "convince" darling-dmg developers (...also us) to add a `DARLING` check to make darling-dmg build & work under Darling.

Reply:
I see. So `#if defined(__APPLE__) && !defined(DARLING)` is the way to go, then.

Reply:
I looked in fuse hoping to find some kind of `__OSXFUSE__`, but no.

Is what you proposed (`#if defined(__APPLE__) && !defined(DARLING)`) ok ?
Or maybe something like `DARLING_OSXFUSE` ? That way, no regression possible when compiling Darling-hq...

Reply:
Yes, it's ok, that's why I've proposed it ;)

A proper way to tell OSXFUSE from FUSE would be better, but if that's not an option, using `DARLING` is always fine.

> Or maybe something like DARLING_OSXFUSE ? That way, no regression possible when compiling Darling-hq...

Come again?